iree/hal/metal/metal_device.mm

```diff
@@ -0,0 +1,194 @@
+    // We use commandBufferWithUnretainedReferences here to be performant. This is okay becasue
+    // IREE tracks the lifetime of various objects with the help from compilers.
+    id<MTLCommandBuffer> cmdbuf = [static_cast<MetalCommandQueue*>(common_queue_)->handle()
+        commandBufferWithUnretainedReferences];
+    command_buffer = MetalCommandBuffer::Create(mode, command_categories, cmdbuf);
+    // TODO: WrapCommandBufferWithValidation(allocator(), std::move(impl));
+    return command_buffer;
+  }
+}
+
+StatusOr<ref_ptr<Event>> MetalDevice::CreateEvent() {
+  IREE_TRACE_SCOPE0("MetalDevice::CreateEvent");
+  return UnimplementedErrorBuilder(IREE_LOC) << "MetalDevice::CreateEvent";
+}
+
+StatusOr<ref_ptr<Semaphore>> MetalDevice::CreateSemaphore(uint64_t initial_value) {
+  IREE_TRACE_SCOPE0("MetalDevice::CreateSemaphore");
+  return MetalSharedEvent::Create(metal_handle_, initial_value);
+}
+
+Status MetalDevice::WaitAllSemaphores(absl::Span<const SemaphoreValue> semaphores,
+                                      Time deadline_ns) {
+  IREE_TRACE_SCOPE0("MetalDevice::WaitAllSemaphores");
+  // Go through all MetalSharedEvents and wait on each of them given we need all of them to be
+  // signaled anyway.
+  for (int i = 0; i < semaphores.size(); ++i) {
+    auto* semaphore = static_cast<MetalSharedEvent*>(semaphores[i].semaphore);
+    IREE_RETURN_IF_ERROR(semaphore->Wait(semaphores[i].value, deadline_ns));
+  }
+  return OkStatus();
+}
+
+StatusOr<int> MetalDevice::WaitAnySemaphore(absl::Span<const SemaphoreValue> semaphores,
+                                            Time deadline_ns) {
+  IREE_TRACE_SCOPE0("MetalDevice::WaitAnySemaphore");
+
+  if (semaphores.empty()) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "expected to have at least one semaphore";
+  }
+
+  // If there is just one semaphore, just wait on it.
+  if (semaphores.size() == 1) {
+    auto* semaphore = static_cast<MetalSharedEvent*>(semaphores[0].semaphore);
+    IREE_RETURN_IF_ERROR(semaphore->Wait(semaphores[0].value, deadline_ns));
+    return 0;
+  }
+
+  // Otherwise, we need to go down a more complicated path by registering listeners to all
+  // MTLSharedEvnts to notify us when at least one of them has done the work on GPU by signaling a
+  // semaphore. The signaling will happen in a new dispatch queue; the current thread will wait on
+  // the semaphore.
+
+  dispatch_time_t timeout = DeadlineToDispatchTime(deadline_ns);
+
+  // Store the handle as a __block variable to allow blocks accessing the same copy for the
+  // semaphore handle on heap.
+  // Use an initial value of zero so that any semaphore signal will unblock the wait.
+  __block dispatch_semaphore_t work_done = dispatch_semaphore_create(0);
+  // Also create a __block variable to store the index for the signaled semaphore.
+  __block int signaled_index = 0;
+
+  dispatch_queue_t wait_notifier =
+      dispatch_queue_create("MetalDevice::WaitAnySemaphore Notifier", NULL);
```

Comment:
Two things here:

1) Should we cache/reuse this queue (or pool queues if needed) instead of creating a new one for each `WaitAnySemaphore` call? I'd expect this to have some overhead that we'd like to avoid (really anything that can run on the hot path which uses strings is a warning flag to me...).
2) https://developer.apple.com/documentation/dispatch/1453030-dispatch_queue_create recommends "a reverse-DNS naming style (com.example.myqueue)". How about `com.google.iree.wait_any_semaphore_notifier`?

That page also says (emphasis mine) "A string label to attach to the queue to **uniquely** identify it in debugging tools such as Instruments, sample, stackshots, and crash reports." but if we create a new queue for each call to `WaitAnySemaphore` then it isn't unique? I guess it shouldn't be a large problem since this call will block on the calling thread and `dispatch_release(wait_notifier);` is called later on, so we won't have multiple active at a time.

Reply:
should definitely cache the queue here

Reply:
Awesome! Thanks for the great comments! Makes sense for me. Was just trying to "make things work", so didn't pay much attention here. :)